physical/raft/transaction.go

```diff
@@ -0,0 +1,605 @@
+		lastKey := ""
+		if len(keys) > 0 {
+			lastKey = keys[len(keys)-1]
+		}
+		var mergedEntries []string
+		for updateEntry := range updates {
+			if updateEntry < entry && updateEntry > lastKey {
+				mergedEntries = append(mergedEntries, updateEntry)
+				delete(updates, updateEntry)
+			}
+		}
+		sort.Strings(mergedEntries)
+		keys = append(keys, mergedEntries...)
+		if len(keys) > 0 {
+			lastKey = keys[len(keys)-1]
+		}
+
+		if isFolder && len(keys) > 0 && lastKey == entry {
+			// This folder was already seen; don't revisit it.
+			continue
+		}
+
+		// Otherwise, include the entry.
+		keys = append(keys, entry)
+		delete(updates, entry)
+	}
+
+	// Finally, attempt to merge newly added entries one more time. This
+	// handles the case when there were no on-disk entries, or when there
+	// were too few and subsequent entries were added here.
+	lastKey := ""
+	if len(keys) > 0 {
+		lastKey = keys[len(keys)-1]
+	}
+	var mergedEntries []string
+	for updateEntry := range updates {
+		if updateEntry > lastKey {
+			mergedEntries = append(mergedEntries, updateEntry)
+			delete(updates, updateEntry)
+		}
+	}
+	sort.Strings(mergedEntries)
+	keys = append(keys, mergedEntries...)
+
+	// We may end up with extra keys as a result of adding all locally
+	// updated ones; if we have too many, trim it down.
+	if limit > 0 && len(keys) > limit {
+		keys = keys[:limit]
+	}
+
+	// Now that we have our result, save the operation in the log
+	// for verification. To do so, we hash the results.
+	listParams, contentsHash, err := createListVerificationEntry(prefix, after, limit, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the read to the verification log, to ensure nobody else has written
+	// to it while the transaction was operating.
+	t.log.Operations = append(t.log.Operations, &LogOperation{
+		OpType: verifyListOp,
+		Key:    listParams,
+		Value:  contentsHash,
+	})
+
+	return keys, nil
+}
+
+func (t *RaftTransaction) Commit(ctx context.Context) error {
+	t.l.Lock()
+	defer t.l.Unlock()
+
+	if t.haveFinishedTx {
+		return nil
+	}
+
+	// The transaction is done; release the permit pool entry now that we're
+	// mostly done with the underlying transaction.
+	//
+	// Also unlock the read lock on the underlying fsm.
+	defer func() {
+		t.b.fsm.l.RUnlock()
+		t.b.txnPermitPool.Release()
+		t.haveFinishedTx = true
+	}()
+
+	// Always rollback the underlying transaction.
+	if err := t.tx.Rollback(); err != nil {
+		return err
+	}
+
+	// If no writes have occurred, we don't need to send a log to Raft. We
+	// might have conflicted on a verification, but we won't negatively impact
+	// an other writer (due to not causing a conflict ourselves). Our state of
+	// reads were guaranteed to be consistent, so it would be no different than
+	// having executed in a read-only transaction.
+	if !t.writable || !t.haveWritten {
+		// Clear our log and updates, restoring us to an empty state.
```

Comment:
This should probably go in the defer; I don't think we'd want to potentially leave this dangling in the event of an error.